keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,480 @@
+for most values:
+
+```golang
+type Foo struct {
+  // +default=32
+  Integer int
+
+  // +default=bar
+  String string
+
+  // +default=[popcorn, chips]
+  StringList []string
+}
+```
+while still allowing more complex expressions:
+```
+// +default=[{port: 80, name: http}]
+```
+
+Forcing the "one-line" subset allows us to sidestep a lot of the YAML
+weirdness in syntax, but we can still just use a YAML parser to parse
+the expressions if we want.
+
+### CRD Normalization
+
+In order to have a consistent behavior between CRDs and built-in, the
+following two things need to happen:
+
+1. Using the flag with CRDs requires the presence of a
+`x-kubernetes-remove-nulls` flag set to true at the root level of the
+object (automatically set when using defaults in built-in types). When
+present, the apiserver will remove all null values from CRDs before
+defaulting, after validation, on all updates (including patches, after
+the patch is applied).
+
+2. Non-pointer structs will, in both kubebuilder and for built-in types,
+disallow the use of `+default` and force a value of `{}`, since this is
+automatically done by Go deserialization:
+```golang
+Type Root struct {
+  // Defaults on non-pointer structs are FORBIDDEN:
+  // +default={name: entry, number: 12}
+  Entry SubLevel
+}
+
+Type SubLevel struct {
+  Name string
+  Number int
+}
+```
+
+### Examples
+
+#### Non-pointer structs
+
+Given the following type, defined either in Kubebuilder or as a built-in type:
+```golang
+type Root struct {
+  Entry SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      $ref: SubLevel
+      default: {}
+SubLevel:
+  type: object
+  properties:
+    name:
+      type: string
+      default: default-name
+    number:
+      type: integer
+      default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+#### Struct pointers
+
+Struct pointers can be have a specific default since they are distinct
+from unspecified non-pointers structs. Modifying the type from the
+previous example:
+```golang
+type Root struct {
+  // This is now a pointer, we can provide a default:
+  // +default={name: pointer-name}
+  Entry *SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
```

Comment:
Is this a good example? Shouldn't we forbid it? I would expect that the field defaults in the top-level default value match those in the fields. We could even go one step further and disallow non-`{}` defaults for structs.

Reply:
+1 to avoid any way of having two different default values for the same field via the +default mechanism. That opens us up to having to define and document rules for which value gets applied when.

Reply:
It's a good example because it shows how things work distinctively.

There is an interesting distinction between: "this is the default item" and "this is the default value for the field", e.g.
```yaml
port:
  default: {number: 53, protocol: UDP}
  properties:
    number:
      type: integer
      # no default
    protocol:
      type: string
      default: TCP
```
This could also apply to a list by the way where the 1 defaulted item would have a UDP port opened and the protocol would still default to TCP.

Reply:
I do like a CRD authors ability to override a default with their preferred default... I am also concerned with the confusion and challenges mentioned by jpbetz

Reply:
I think the feature is useful and important, we should let people build what is useful for them. We can maybe avoid the foot-gun by either:
- Have a warning when a default is specified on a non-pointer struct,
- Be smarter and have a warning when that default doesn't match defaults in nested fields

Reply:
> I do like a CRD authors ability to override a default with their preferred default... 

I don't think this is the purpose of this. For your usecase kubebuilder better has a way to override. In our API repo https://github.com/openshift/api/blob/master/operator/v1/0000_20_kube-apiserver-operator_01_config.crd.yaml-patch we allow to patch whatever kubebuilder derives from types. I would love to see this implemented by kubebuilder natively, something like:
```
// +kubebuilder:schema-override:containers[].name={pattern: ...}
```
In other words, overrides do not need support in the CRD schema API, but only in tooling, here in controller-gen.

Reply:
Limiting the defaults of complex types (objects and arrays) to single-level values (`{}` and `[]`) would make generating the defaulting functions significantly easier.

Reply:
Yeah, I'm really not convinced that it will change the complexity of the code, and I can already see use-cases where this makes sense, I'm sure users will find others.

Reply:
I think we agree that this is a good behavior.